src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,420 @@
+         * piece-meal or if part of the IKM is to be supplied by a hardware
+         * crypto device. This method appends to the existing list of values or
+         * creates a new list if there are none yet.
+         * <p>
+         * This supports the use-case where a label can be applied to the IKM
+         * but the actual value of the IKM is not yet available.
+         *
+         * @param ikm
+         *     the initial key material value (null or empty values will not be added)
+         *
+         * @return a new {@code Extract} object
+         */
+        public Builder addIKM(byte[] ikm) {
+            if (ikm != null && ikm.length != 0) {
+                return addIKM(new SecretKeySpec(ikm, "Generic"));
+            } else {
+                return this;
+            }
+        }
+
+        /**
+         * {@code addSalt} may be called when the salt value is to be assembled
+         * piece-meal or if part of the salt is to be supplied by a hardware
+         * crypto device. This method appends to the existing list of values or
+         * creates a new list if there are none yet.
+         * <p>
+         * This supports the use-case where a label can be applied to the salt
+         * but the actual value of the salt is not yet available.
+         *
+         * @param salt
+         *     the salt value (null values will not be added)
+         *
+         * @return a new {@code Extract} object
+         */
+        public Builder addSalt(SecretKey salt) {
+            if (salt != null) {
+                salts.add(salt);
+            }
+            return this;
+        }
+
+        /**
+         * {@code addSalt} may be called when the salt value is to be assembled
+         * piece-meal or if part of the salt is to be supplied by a hardware
+         * crypto device. This method appends to the existing list of values or
+         * creates a new list if there are none yet.
+         * <p>
+         * This supports the use-case where a label can be applied to the salt
+         * but the actual value of the salt is not yet available.
+         *
+         * @param salt
+         *     the salt value (null or empty values will not be added)
+         *
+         * @return a new {@code Extract} object
+         */
+        public Builder addSalt(byte[] salt) {
+            if (salt != null && salt.length != 0) {
+                return addSalt(new SecretKeySpec(salt, "Generic"));
+            } else {
+                return this;
+            }
+        }
+    }
+
+    /**
+     * Static helper-method that may be used to initialize a {@code Builder}
+     * with an empty {@code Extract}
+     * <p>
+     * Note: one or more of the methods {@code addIKM} or {@code addSalt} should
+     * be called next, before calling build methods, such as
+     * {@code Builder.extractOnly()}
+     *
+     * @return a {@code Builder} to mutate
+     */
+    static Builder extract() {
```

Comment:
The expand method return an Expand object, the extractExpand returns a ExtractExpand. It looks a bit strange to have the extract method return a Builder rather than extract something or return an Extract object. Try renaming it to builder and see if it looks any better.

Reply:
I agree with Alan the name is a little confusing because it doesn't directly return an `Extract` object. Suggest renaming to `builder()` as Alan suggests.